java/src/main/java/org/digma/intellij/plugin/idea/psi/java/AbsJaxrsFramework.java

```diff
@@ -114,59 +112,102 @@ private boolean isJaxRsHttpRelevant() {
+     */
     @Override
-    public void endpointDiscovery(@NotNull PsiFile psiFile, @NotNull DocumentInfo documentInfo) {
+    public List<EndpointInfo> lookForEndpoints(@NotNull PsiFile psiFile) {
         lateInit();
         if (!isJaxRsHttpRelevant()) {
-            return;
+            return Collections.emptyList();
         }
 
-        var psiFacade = JavaPsiFacade.getInstance(project);
-        final PsiClass appPathAnnotationClass = psiFacade.findClass(getApplicationPathAnnotationClassFqn(), GlobalSearchScope.allScope(project));
-
-        List<PsiClass> allClassesInFile = JavaPsiUtils.getClassesWithin(psiFile);
-        for (PsiClass currClass : allClassesInFile) {
-            final PsiAnnotation controllerPathAnnotation = JavaPsiUtils.findNearestAnnotation(currClass, JAX_RS_PATH_ANNOTATION_STR());
+        Collection<PsiClass> allClassesInFile = PsiTreeUtil.findChildrenOfType(psiFile, PsiClass.class);
 
-            List<PsiMethod> methodsInClass = Arrays.asList(currClass.getMethods());
-            for (PsiMethod currPsiMethod : methodsInClass) {
+        Set<PsiMethod> candidateMethods = new HashSet<>();
+        for (final PsiClass currClass : allClassesInFile) {
+            List<PsiMethod> methodsInClass = JavaPsiUtils.getMethodsOf(currClass);
+            for (final PsiMethod currPsiMethod : methodsInClass) {
                 final PsiAnnotation methodPathAnnotation = JavaPsiUtils.findNearestAnnotation(currPsiMethod, JAX_RS_PATH_ANNOTATION_STR());
-                if (methodPathAnnotation == null && controllerPathAnnotation == null) {
-                    continue; // skip since could not find annotation of @Path, in either class and or method
+                boolean hasPath = methodPathAnnotation != null;
+                boolean hasHttpMethod = httpMethodsAnnotations.stream()
+                        .anyMatch(it -> JavaPsiUtils.findNearestAnnotation(currPsiMethod, it.getClassNameFqn()) != null);
+
+                if (hasPath || hasHttpMethod) {
+                    candidateMethods.add(currPsiMethod);
                 }
+            }
+        }
 
-                Set<String> appPaths = evaluateApplicationPaths(appPathAnnotationClass, currClass);
+        return handleCandidateMethods(candidateMethods).stream().toList();
+    }
 
-                for (JavaAnnotation currExpectedAnnotation : httpMethodsAnnotations) {
-                    PsiAnnotation httpMethodAnnotation = JavaPsiUtils.findNearestAnnotation(currPsiMethod, currExpectedAnnotation.getClassNameFqn());
-                    if (httpMethodAnnotation == null) {
-                        continue; // skipping since could not find annotation of HTTP Method, such as @GET
-                    }
-                    String endpointSuffixUri = combinePaths(controllerPathAnnotation, methodPathAnnotation);
+    @Override
+    public List<EndpointInfo> lookForEndpoints(@NotNull SearchScope searchScope) {
+        lateInit();
```

Comment:
what is the purpose of lateInit ? why not do it in the constructor?

Reply:
constructor might be too early

Reply:
Yes, and i think i know why. because when constructing the instance indexes are not ready yet and the instance requires indexes to be ready because its using JavaPsiFacade search methods.
It starts because all the framework classes are constructed as members on JavaLanguageService which is constructed before indexes are ready. JavaLanguageService does not require indexes to be ready.

I think that all the framework classes don't need to be instances of JavaLanguageService, they can be constructed when needed and usually its in a stage when indexes are ready. their construction is not heavy so an instance can be constructed every time its needed.

fro example JavaEndpointNavigationProvider.buildEndpointAnnotations is called in smart mode so indexes are ready, and you can construct the EndpointDiscovery only then and avoid the lateInit.
or it can all be static methods like JavaSpanDiscoveryUtils for example.

as i see all these framework classes don't have state, they do have members , for example 'PsiClass controllerAnnotationClass'  or 'PsiClass jaxrsPathAnnotationClass'
but this is also not the best, jetbrains recommend not keeping instances of PSI elements as members because PSI elements may be disposed and new ones built. you never know when this PSI element becomes invalid because of reparse.  if you really need to keep an reference to PSI elements you should keep it in SmartPsiElementPointer. 

So i think all the framework classes can be static methods. search for annotations or classes when you need them, make sure its always run in smart mode with read action